File: Event.ts
import { EventDef } from "./EventDef";

export class Event extends EventDef {
  constructor(eventId?: string, eventName?: string, eventDate?: string, maxParticipants?: number) {
    super();
    this.setEventId(eventId                || "");
    this.setEventName(eventName            || "");
    this.setEventDate(eventDate            || "");
    this.setMaxParticipants(maxParticipants || 0);
  }

  getEventId(): string               { return this.eventId; }
  getEventName(): string             { return this.eventName; }
  getEventDate(): string             { return this.eventDate; }
  getMaxParticipants(): number       { return this.maxParticipants; }
  getParticipantsRegistered(): number { return this.participantsRegistered; }

  setEventId(eventId: string): void         { this.eventId         = eventId; }
  setEventName(eventName: string): void     { this.eventName       = eventName; }
  setEventDate(eventDate: string): void     { this.eventDate       = eventDate; }
  setMaxParticipants(max: number): void     { this.maxParticipants = max; }
}

File: Event.cpp
#include "Event.h"
#include "Student.h"
#include <iostream>
using namespace std;

// Constructor
Event::Event(string id, string name, string date, int maxP) {
    eventID                = id;
    eventName              = name;
    eventDate              = date;
    maxParticipants        = maxP;
    participantsRegistered = 0;
    cout << "[Event Created]: " << eventName << "\n";
}

// Destructor
Event::~Event() {
    cout << "[Event Removed]: " << eventName << "\n";
}

// Getters
string Event::getEventID()                { return eventID; }
string Event::getEventName()              { return eventName; }
string Event::getEventDate()              { return eventDate; }
int    Event::getMaxParticipants()        { return maxParticipants; }
int    Event::getParticipantsRegistered() { return participantsRegistered; }

// Setters
void Event::setEventID(string id)       { eventID         = id; }
void Event::setEventName(string name)   { eventName       = name; }
void Event::setEventDate(string date)   { eventDate       = date; }
void Event::setMaxParticipants(int max) { maxParticipants = max; }

// Methods
bool Event::isAvailable() {
    return participantsRegistered < maxParticipants;
}

void Event::registerStudent(Student &student) {
    if (isAvailable()) {
        participantsRegistered++;
        cout << "  Success! " << student.getFirstName() << " "
             << student.getLastName() << " registered for '" << eventName << "'.\n";
    } else {
        cout << "  Sorry, '" << eventName << "' is full.\n";
    }
}

void Event::displayInfo() {
    string status = isAvailable() ? "Available" : "Full";
    cout << "  Event ID : " << eventID << "\n";
    cout << "  Name     : " << eventName << "\n";
    cout << "  Date     : " << eventDate << "\n";
    cout << "  Slots    : " << participantsRegistered << " / " << maxParticipants << "\n";
    cout << "  Status   : " << status << "\n";
}